corehq/apps/reports/standard/deployments.py

```diff
@@ -317,7 +317,7 @@ def user_locations_dict(self, users):
             for loc_id in user.get('assigned_location_ids'):
```

Comment:
How is the populated on the user though? Can you confirm if this is actually available only for mobile users? Would be good to test this locally for a web user as well.

Reply:
This is an available field on the `CouchUser` model ([code ref](https://github.com/dimagi/commcare-hq/blob/50417890a5c241c062f1aa30a9f9a132d48ccd00/corehq/apps/users/models.py#L991)), so it's available for both mobile and web users. But looking through the code, it seems this field only gets used for `CommCareUser` (blank for web users). For web users, their locations are being added [here](https://github.com/dimagi/commcare-hq/blob/50417890a5c241c062f1aa30a9f9a132d48ccd00/corehq/apps/users/models.py#L2120) under a domain membership instance.

Good catch, I'll need to make a code change to accommodate this. To make things easier, we could have a helper function to correctly wrap the user doc. In code, this might look something like:
```
def correctly_wrap_user(self, user_doc):
    from corehq.apps.users.models import CouchUser, WebUser, CommCareUser
    couch_user = CouchUser.wrap(user_doc)
    if couch_user.is_web_user():
        return WebUser.wrap(user_doc)
    return CommCareUser.wrap(user_doc)
```
Once we have the wrapped user, we can simply call `get_location_ids()` for assigned locations or `get_location_id()` for the primary location on either mobile or web users where we need the location IDs.

@mkangia does the above implementation sound fine to you?

Reply:
FYI there's also a `CouchUser.wrap_correctly()` that does what you want here.

Reply:
Oh that's perfect, thanks Ethan.

Reply:
I see we are getting users from elasticsearch here.
I am still not sure what value assigned_location_ids would have for a web user. I think if you can check an ES doc for a web user and for a mobile user, it would be clear how we can fetch location ids just for the relevant domain.
Limiting to domain in sql query seems like a good step to add as well.

Reply:
@mkangia For ES, the assigned locations for a web user are stored [here](https://github.com/dimagi/commcare-hq/blob/35379ef54421e9271ab6b0126b9067295b936b50/corehq/apps/es/mappings/user_mapping.py#L125) in the `domain_memberships` field. If we just use the ES doc then we'd have to loop through all the user's domain memberships to find the right domain so that we can retrieve the user's list of assigned locations. Wrapping the users as a `WebUser` or `CommCareUser` instance instead allows us to simplify this process and simply call `get_location_ids()` to retrieve assigned locations for the user based on the domain.

Reply:
I have pushed a commit with the above proposed changes ([581bfe1](https://github.com/dimagi/commcare-hq/pull/34744/commits/581bfe1a7dd92eb5d6282dc48f37f4599eeb899b)). @mkangia Please let me know if the implemented approach makes sense, happy to answer any questions you may have.

Reply:
Thanks the commit @zandre-eng , useful to understand the change.

I understand that using `wrap_correctly` gives you access to `get_location_ids`, so it makes things easier programatically, though since we already had that information in the elasticsearch doc, doing a wrap operation just to get access to a function instead of fetching it directly seems an overkill.
I am also not very well familiar with locations in users es docs. let me ask Ethan

Reply:
Hey @esoergel 

Just checking if you have any recommendations on this.

we are getting a list of user ES docs and want their assigned locations for the current request domain. Should we be able to fetch it from ES docs itself or its better to wrap it to couch user first?

Reply:
I think you should be able to get it from the unwrapped docs too, though if web users are in the mix, you'll have to pull it from the correct domain membership for them

Reply:
thanks @esoergel for your inputs here.

It does seem better to use `wrap_correctly` then. I thought about this a bit more and I prefer using existing code to do this correctly rather than fiddling around with elasticsearch doc unless that is already done somewhere (which I could not find in HQ anywhere).

Reply:
Aside: The way we do location assignment as a field on the couch user is a frequent source of frustration (especially since we store it in different places between web and mobile users).  One of these days I'd like to move it to SQL, something like:
```python
class HqUser(models.Model):
    django_user = models.OneToOneField(User)
    locations = models.ManyToManyField(
        SQLLocation, through=`LocationAssignment`, related_name='users')

class LocationAssignment(models.Model):
    domain = models.CharField()
    hq_user = models.ForeignKey(HqUser)
    location = models.ForeignKey(SQLLocation)
    is_primary = models.BooleanField()
```
Then we could finally work with users and locations in bulk using joins

Reply:
thanks @esoergel 

That would be the SQL way of doing it.

I do think however if, 
1. primary location id could be saved on user itself? is that information duplication? possibly.
2. in order to look up locations of a certain set of users, we are going to join three tables. I wonder if in terms of SQL that would hinder the performance considering that is quite often.

I don't know if keeping a list column that has list of location ids on the HQUser would be a bad idea.
Though that does making fetching locations for users in bulk impossible.

So,
For a single user we would still fetch by ID though for bulk we would do a join. Seems good to me.

That is going to be some migration.